File: mqm.pl
use warnings;
use strict;

my $root={};
my $time = time;
open (D, "<", 'dict');
while (my $text = <D>){
	chomp $text;
	my $hashRef = \$root;    
	my @tmp = split(' ', $text);
	foreach my $tmp (@tmp){
		$hashRef = \$$hashRef->{$tmp};
		if (\$tmp==\$tmp[-1]){
			$hashRef = \$$hashRef->{''};
			$$hashRef=1;
		}
	}
	last if $n>5;
}

warn time - $time;

sub found_match {
	my ($fm_print_line, $fm_match, $fm_found, $fm_ref)= @_;
	$fm_print_line .= "<FOUND>".$fm_match."<\/FOUND> ";
	$fm_found =1;
	$fm_ref = \$root;
	$fm_match ="";	
	return ($fm_print_line, $fm_match, $fm_found, $fm_ref);
}

open (S, "<", 'text');
$time = time;
while (my $line = <S>){
	my $found =0;
	chomp $line;
	my $hashRef = \$root;
	my $switch=0;
	my @tmp = split(' ', $line);
	my $match="";
	my $print ="";
	for (my $i=0; $i<=$#tmp; $i++){
		my $tmp = $tmp[$i];
		if (defined $$hashRef->{$tmp}){
			$hashRef=\$$hashRef->{$tmp};
			$match.=$tmp." ";
			if (defined $$hashRef->{''}){
				$match=~s/ $//;
				($print, $match, $found, $hashRef) = found_match($print, $match, $found, $hashRef);
			}
		}
		else {
			$hashRef=\$root;
			if (not defined $$hashRef->{$tmp}){
				$print.=$match.$tmp." ";
			}
			else {
				$print.=$match;
				$match="";
				$hashRef = \$$hashRef->{$tmp};
				$match.=$tmp." ";
				if (defined $$hashRef->{''}){
					$match =~s/ $//;
					($print, $match, $found, $hashRef) = found_match($print, $match, $found, $hashRef);
				}
			}
		}
	}
	$print=~s/ $//;
	print $print, "\n" if $found ==1;
}
warn time - $time;

File: mqm.py
dict_file = open("dict", 'r')
dict = {}

for dict_entry in dict_file:
    ref = dict
    tmp = dict_entry.rstrip('\n').split(' ')
    for i in tmp:
        if i not in ref:
            ref[i]={}
        ref = ref[i]
        if (i == tmp[-1]):
            ref['']=1
# print(dict)

def found_match(fm_print_line, fm_match, fm_found, fm_ref):
	fm_print_line = fm_print_line+"<FOUND>"+fm_match.rstrip(' ')+"<\/FOUND> "
	fm_found =1
	fm_ref = dict
	fm_match =""
	return fm_print_line, fm_match, fm_found, fm_ref

text_file = open("text", 'r')
for line in text_file:
	found =0
	ref = dict
	switch =0
	arr = line.rstrip('\n').split(' ')
	match =""
	print_line = ""
	for i in range(len(arr)):
		tmp = arr[i]
		if (tmp in ref):
			ref = ref[tmp]
			match=match+tmp+' '
			if ('' in ref):
				print_line, match, found, ref = found_match(print_line, match, found, ref)
		else:
			ref = dict
			if (tmp not in ref):
				print_line=print_line+match+tmp+' '
			else:
				print_line=print_line + match
				match =''
				ref = ref[tmp]
				match=tmp+' '
				if ('' in ref):
					print_line, match, found, ref = found_match(print_line, match, found, ref)

	if found ==1:
		print(print_line.rstrip(' '))
